semantic/src/main/kotlin/tools/samt/semantic/SemanticModel.kt

```diff
@@ -0,0 +1,284 @@
+                            }
+                        }
+
+                        is WildcardImportNode -> {
+                            // Import all types from the package
+                            val type = global.resolveType(import.name)
+                            if (type != null) {
+                                if (type is PackageType) {
+                                    type.sourcePackage.types.forEach { (name, type) ->
+                                        putIfAbsent(name, type)
+                                    }
+                                } else {
+                                    controller.createContext(file.sourceFile).error {
+                                        message("Import '${import.name}.*' must point to a package and not a type")
+                                        highlight(
+                                            "illegal wildcard import", import.location, suggestChange = "import ${
+                                                import.name.components.dropLast(1).joinToString(".") { it.name }
+                                            }.*")
+                                    }
+                                }
+                            } else {
+                                controller.createContext(file.sourceFile).error {
+                                    message("Import '${import.name}.*' could not be resolved")
+                                    highlight("unresolved import", import.location)
+                                }
+                            }
+                        }
+                    }
+                }
+
+                // Add all top-level packages to scope
+                topLevelPackages.forEach { putIfAbsent(it.packageName, it) }
+
+                // Add built-in types
+                putIfAbsent("Int", IntType)
+                putIfAbsent("Float", FloatType)
+                putIfAbsent("String", StringType)
+                putIfAbsent("Boolean", BooleanType)
+            }
+
+            file.sourceFile to typeLookup
+        }
+
+        global.allSubPackages.forEach { subPackage ->
+            subPackage.records.forEach { record ->
+                val typeLookup = typeLookupBySource.getValue(record.definition.location.source)
+                record.fields.forEach { field ->
+                    field.type = resolveTypeReference(field.type, subPackage, typeLookup)
+                    // TODO all other types
+                }
+            }
+        }
+
+        // TODO: detect cycles
+
+        return global
+    }
+
+    private fun resolveTypeReference(
+        reference: TypeReference,
+        parentPackage: Package,
+        typeLookup: Map<String, Type>,
+    ): TypeReference {
+        if (reference !is UnresolvedTypeReference) {
+            return reference
+        }
+
+        val exp = reference.expression
+
+        // TODO parse type expression, this is dummy code
+        if (exp is IdentifierNode) {
```

Comment:
use a when statement

Reply:
Code will be re-written entirely, also using when-statements and other nice features 😉